### DISCUSSION TITLE
Question Regarding User Creation Flow in Supabase and Database Synchronization

### DISCUSSION THREAD

**Discussion by saurabhchalke** [gokulkrishh/expense.fyi#93] [Q&A] [answered]
Hello team,

I've been exploring the project and trying to set it up locally. I've come across a behavior I'd like to clarify concerning the user creation process, especially in relation to Supabase and the database synchronization.

When I attempt to sign up a user, I observed the following:

1. A magic link is generated and an email is sent to the user for authentication.
2. Upon checking, a new authentication user is created in Supabase's authentication system.
3. However, this new user is not being persisted in the `users` table in the database, even though the user model is defined in the `schema.prisma`.

Given the above, I have a few questions:

1. **Supabase Triggers/Functions**: Does the project utilize any Supabase triggers or functions to automatically synchronize or create entries in the `users` table based on authentication events? If so, could you provide some insights or pointers on how this is set up?

2. **Local Development**: Are there specific steps or configurations needed for local development to ensure that the Supabase authentication and database synchronization work seamlessly? I've ensured my local Supabase configurations are set correctly, but I might be missing something.

3. **Additional Mechanisms**: Apart from the above, are there any other mechanisms or processes in place that handle the synchronization between the Supabase auth system and the `users` table?

I appreciate the work you've put into this project, and I'm eager to understand it more deeply. Any guidance or insights would be immensely helpful.

Thank you for your time!

**Comment by gokulkrishh** [comment, marked as answer]
Hi @saurabhchalke Glad you liked this app (Do support by giving a star) and i will try to answer your questions below.  

1. `Supabase Triggers/Function` - Yes you need to setup a trigger function in supabase.[ This is the trigger function ](https://github.com/gokulkrishh/expense.fyi/blob/main/prisma/functions/functions.sql) i am using. You can follow the docs in supabase on how to set it up.
  > So basically, this function will take the signup users and put them into users table.
2. `Local Development:` Yes you need to setup ENV keys into `.env`file to run the app in your local. You can find required keys in [example.env](https://github.com/gokulkrishh/expense.fyi/blob/main/.env.example).  
   1. Supabase
   2. Resend
   3. [prisma-field-encryption](https://github.com/47ng/prisma-field-encryption) for encryption.
   These should be enough to get the app up and running in local
3. `Additional Mechanisms` - None. Once the user data is put into the users table, from there app update the table for users from app such as locale, CRUD for data, deleting the user and their data etc.

If you follow the above steps, you should be able to run the app locally. Let me know if you have any more questions.

**Comment by saurabhchalke** [comment]
Thank you for the detailed response. I've followed your guidance and managed to set things up successfully. Specifically, I created the following trigger + function on Supabase to handle the user creation process:

```sql
/**
* This trigger automatically creates a user entry when a new user signs up via Supabase Auth.
*/ 
create function public.create_user_on_auth() 
returns trigger as $$
begin
  insert into public.users (id, email)
  values (new.id, new.email);
  return new;
end;
$$ language plpgsql security definer;

create trigger create_new_user
  after insert on auth.users
  for each row execute procedure public.create_user_on_auth();
```

This did the trick for me and everything's working as expected now. Thank you once again, and I'm looking forward to contributing and learning more from this project.

**Reply by gokulkrishh** [reply]
Cool!
